### PR TITLE
Add gevent worker class option/Decouple Gunicorn config

### PR DESCRIPTION
**Pull request by norenstrakh-r7**
### How Has This Been Tested?

Built several plugins(dig|csv|vulndb|whois) and deployed as cloud plugin in staging environment
the 

### Types of changes
- this PR adds an option to set the worker class(sync|gevent)
- it decouples GUNICRON config from the application, experimenting Gunicorn changes won't require code changes 

### Checklist:

- [ v] I have updated the documentation accordingly (if changes are required).
- [ v] I have tested the changes locally.

### Miscellaneous Details:

This PR requires icon-plugin changes to update generation of the `bin/komand_${PLUGIN_NANE}` to support early monkey patching with gevent

### CODE REVIEW

**Review by tschmidtke-r7 — DISMISSED** [review]
This is exciting! Personally I'd prefer the config file to be `GUNICORN_CONFIG_FILE`, but that's more of a nitpick.

File: samples/example/requirements.txt [hunk 1]
```diff
@@ -1,3 +1,4 @@
 # List third-party dependencies here, separated by newlines. 
 # All dependencies must be version-pinned, eg. requests==1.2.0
-# See: https://pip.pypa.io/en/stable/user_guide/#requirements-files
\ No newline at end of file
+# See: https://pip.pypa.io/en/stable/user_guide/#requirements-files
+gunicorn[gevent]
```
Comment (norenstrakh-r7) [inline samples/example/requirements.txt:4]: can be removed

File: dockerfiles/3-38-slim [hunk 2]
```diff
@@ -4,12 +4,17 @@ LABEL organization=rapid7 \
...
 			type=plugin
 
+RUN apk update
+RUN apk add build-base libffi-dev openssl-dev
+
 ENV SSL_CERT_FILE=/etc/ssl/certs/ca-certificates.crt \
 		SSL_CERT_DIR=/etc/ssl/certs \
 		REQUESTS_CA_BUNDLE=/etc/ssl/certs/ca-certificates.crt
 
+
```
Comment (zyoutz-r7) [inline dockerfiles/3-38-slim:14]:
```suggestion
```

File: dockerfiles/3-38-slim [hunk 3]
```diff
@@ -4,12 +4,17 @@ LABEL organization=rapid7 \
 			sdk=python \
 			type=plugin
 
+RUN apk update
```
Comment (zyoutz-r7) [inline dockerfiles/3-38-slim:7]: Can we get a comment added here for why the dependencies are necessary

File: insightconnect_plugin_runtime/server.py [hunk 4]
```diff
@@ -38,13 +39,24 @@ class PluginServer(gunicorn.app.base.BaseApplication):
...
+    def __init__(self, plugin, port=10001, workers=1, threads=4, debug=False,
+                 worker_class='sync', worker_connections=200):
+
+        if os.environ.get("GUNICORN_CFG_FILE"):
+            with open(os.environ.get("GUNICORN_CFG_FILE")) as gf:
+                self.gunicorn_config = json.load(gf)
+        else:
+            self.gunicorn_config = {
+                "bind": "%s:%s" % ("0.0.0.0", port),
+                "workers": workers,
```
Comment (zyoutz-r7) [inline insightconnect_plugin_runtime/server.py:51]: Are there any unintended side effects  if someone were to set workers > 1 along with `gevent` class?

File: insightconnect_plugin_runtime/server.py [hunk 5]
```diff
@@ -38,13 +39,24 @@ class PluginServer(gunicorn.app.base.BaseApplication):
...
-            "bind": "%s:%s" % ("0.0.0.0", port),
-            "workers": workers,
-            "threads": threads,
-            "loglevel": "debug" if debug else "info",
-        }
+    def __init__(self, plugin, port=10001, workers=1, threads=4, debug=False,
+                 worker_class='sync', worker_connections=200):
+
+        if os.environ.get("GUNICORN_CFG_FILE"):
+            with open(os.environ.get("GUNICORN_CFG_FILE")) as gf:
```
Comment (zyoutz-r7) [inline insightconnect_plugin_runtime/server.py:46]: Love this addition! Will make it that much easier to investigate other settings that are made available 🔥

File: samples/example/bin/icon_example [hunk 6]
```diff
@@ -1,33 +1,47 @@
...
 
-        self.add_action(actions.SayGoodbye())
+    if os.environ.get("GUNICORN_CFG_FILE"):
+        with open(os.environ.get("GUNICORN_CFG_FILE")) as gf:
+            gunicorn_cfg = json.load(gf)
+            if gunicorn_cfg.get("worker_class", "sync") == "gevent":
+                from gevent import monkey
+                monkey.patch_all()
+    elif 'http' in argv and 'gevent' in argv and not os.environ.get("GUNICORN_CFG_FILE"):
+        from gevent import monkey
```
Comment (zyoutz-r7) [inline samples/example/bin/icon_example:22]: More for icon-plugin changes but can we make sure a comment is added here around why this order of imports is necessary?

File: samples/example/bin/icon_example [hunk 7]
```diff
@@ -1,33 +1,47 @@
...
+def main():
 
-        self.add_action(actions.SayGoodbye())
+    if os.environ.get("GUNICORN_CFG_FILE"):
+        with open(os.environ.get("GUNICORN_CFG_FILE")) as gf:
+            gunicorn_cfg = json.load(gf)
+            if gunicorn_cfg.get("worker_class", "sync") == "gevent":
+                from gevent import monkey
+                monkey.patch_all()
+    elif 'http' in argv and 'gevent' in argv and not os.environ.get("GUNICORN_CFG_FILE"):
```
Comment (zyoutz-r7) [inline samples/example/bin/icon_example:21]: Do we need the addition `and not` condition for this? Could we ever hit this?

File: samples/example/requirements.txt (hunk 1 above)
Comment (zyoutz-r7) [inline samples/example/requirements.txt:4]: Already part of setup.py for runtime.  Not needed for each plugin
```suggestion
```

**Review by zyoutz-r7 — CHANGES_REQUESTED** [review]
Looking good! Couple suggestions

File: dockerfiles/3-38-slim [hunk 8]
```diff
@@ -4,12 +4,17 @@ LABEL organization=rapid7 \
 			sdk=python \
```
Comment (zyoutz-r7) [inline dockerfiles/3-38-slim:4]: And just for confirmation, no changes necessary to the 3-38 (ubuntu) docker file?

File: insightconnect_plugin_runtime/server.py (hunk 4 above)
Comment (norenstrakh-r7) [reply insightconnect_plugin_runtime/server.py:51]: no , there is not, gunicorn will fork another worker, I actually tried it.

File: samples/example/bin/icon_example (hunk 7 above)
Comment (norenstrakh-r7) [reply samples/example/bin/icon_example:21]: let me change this condition,  I think there is a room for improvement

File: dockerfiles/3-38-slim (hunk 8 above)
Comment (norenstrakh-r7) [reply dockerfiles/3-38-slim:4]: yep,  I was able to build `whois` (which we already ranging as cloud-plugin in staging)  without making changes to the base image.

**Comment by norenstrakh-r7** [comment]
> This is exciting! Personally I'd prefer the config file to be `GUNICORN_CONFIG_FILE`, but that's more of a nitpick.

changed

File: dockerfiles/3-38-slim [hunk 9]
```diff
@@ -4,6 +4,10 @@ LABEL organization=rapid7 \
...
 			type=plugin
 
+RUN apk update
+RUN apk add build-base libffi-dev openssl-dev
+
+# Support Gevent dependencies
```
Comment (zyoutz-r7) [inline dockerfiles/3-38-slim:10]: Not sure about ordering of comment
```suggestion
# Support Gevent dependencies
RUN apk update
RUN apk add build-base libffi-dev openssl-dev

```

File: insightconnect_plugin_runtime/server.py (hunk 4 above)
Comment (zyoutz-r7) [reply insightconnect_plugin_runtime/server.py:51]: 👍 perfect

File: samples/example/bin/icon_example [hunk 10]
```diff
@@ -1,33 +1,58 @@
...
-        self.add_trigger(triggers.EmitGreeting())
-
-        self.add_action(actions.SayGoodbye())
-
-        self.add_action(actions.SayHello())
-
-
 def main():
-    """Run plugin"""
+    # If the worker is configured to run as gevent class
```
Comment (zyoutz-r7) [inline samples/example/bin/icon_example:14]: Thank you

File: samples/example/bin/icon_example [hunk 11]
```diff
@@ -1,33 +1,58 @@
...
+    #  and only then importing the rest of the dependencies. It only has to be done
+    #  if the plugin is running in Server mode(http).
+    if 'http' in argv:
+        if os.environ.get("GUNICORN_CONFIG_FILE"):
+            with open(os.environ.get("GUNICORN_CONFIG_FILE")) as gf:
+                gunicorn_cfg = json.load(gf)
+                if gunicorn_cfg.get("worker_class", "sync") == "gevent":
+                    from gevent import monkey
+                    monkey.patch_all()
+        elif 'gevent' in argv:
```
Comment (zyoutz-r7) [inline samples/example/bin/icon_example:31]: 👍

File: README.md [hunk 12]
```diff
@@ -68,6 +68,12 @@ The plugin will be started in `http` mode and listening at `http:0.0.0.0:10001`:
 [2020-02-13 23:21:13 -0500] [56571] [INFO] Booting worker with pid: 56571
 ```
 
+To override Gunicorn config export environment variable GUNICORN_CFG_FILE pointing to json file with config
+See example in samples/example/gcfg.json 
+```
+export GUNICORN_CFG_FILE="./gcfg.json"
```
Comment (zyoutz-r7) [inline README.md:74]:
```suggestion
export GUNICORN_CONFIG_FILE ="./gcfg.json"
```

File: README.md [hunk 13]
```diff
@@ -68,6 +68,12 @@ The plugin will be started in `http` mode and listening at `http:0.0.0.0:10001`:
 [2020-02-13 23:21:13 -0500] [56571] [INFO] Booting worker with pid: 56571
 ```
 
+To override Gunicorn config export environment variable GUNICORN_CFG_FILE pointing to json file with config
```
Comment (zyoutz-r7) [inline README.md:71]:
```suggestion
To override Gunicorn config export environment variable GUNICORN_CONFIG_FILE pointing to json file with config
```

**Review by zyoutz-r7 — CHANGES_REQUESTED** [review]
Couple small suggestions

**Review by zyoutz-r7 — APPROVED** [review]
Love it! Lets make sure @mrinehart-r7 and @gusingh-r7 also get a chance to give there feedback.

**Review by gusingh-r7 — DISMISSED** [review]

**Review by lmilby-r7 — COMMENTED** [review]
Are we looking at including tests for these changes?

**Review by jmcadams-r7 — APPROVED** [review]

**Review by zyoutz-r7 — APPROVED** [review]

**Review by gusingh-r7 — APPROVED** [review]

**Pull request merged**